File: leetcode/binary_search/1891_Cutting_Ribbons/codes/bs.c
#include <stdio.h>
#include <stdlib.h>

int cmp(const void* a, const void* b) {
    return (*(int*)a - *(int*)b);
}

int maxLength(int* ribbons, int ribbonsSize, int k) {
    int maxVal = 0;
    for (int i = 0; i < ribbonsSize; i++)
        if (ribbons[i] > maxVal) maxVal = ribbons[i];
    int left = 1, right = maxVal, ans = 0;
    while (left <= right) {
        int mid = left + (right - left) / 2;
        long long count = 0;
        for (int i = 0; i < ribbonsSize; i++)
            count += ribbons[i] / mid;
        if (count >= k) {
            ans = mid;
            left = mid + 1;
        } else {
            right = mid - 1;
        }
    }
    return ans;
}

File: leetcode/binary_search/1891_Cutting_Ribbons/codes/bs.py
from typing import List

class Solution:
    def maxLength(self, ribbons: List[int], k: int) -> int:
        left, right = 1, max(ribbons)
        ans = 0
        while left <= right:
            mid = (left + right) // 2
            count = sum(r // mid for r in ribbons)
            if count >= k:
                ans = mid
                left = mid + 1
            else:
                right = mid - 1
        return ans
